src/v/cluster/partition.cc

```diff
@@ -390,9 +408,69 @@ partition::timequery(storage::timequery_config cfg) {
     co_return result;
 }
 
+void partition::maybe_construct_archiver() {
+    if (
+      config::shard_local_cfg().cloud_storage_enabled()
+      && _cloud_storage_api.local_is_initialized()
+      && _raft->ntp().ns == model::kafka_namespace
+      && _raft->log().config().is_archival_enabled()) {
+        _archiver = std::make_unique<archival::ntp_archiver>(
+          log().config(), _archival_conf, _cloud_storage_api.local(), *this);
+    }
+}
 ss::future<> partition::update_configuration(topic_properties properties) {
-    co_await _raft->log().update_configuration(
-      properties.get_ntp_cfg_overrides());
+    auto& old_ntp_config = _raft->log().config();
+    auto new_ntp_config = properties.get_ntp_cfg_overrides();
+
+    // Before applying change, consider whether it changes cloud storage mode
+    bool cloud_storage_changed = false;
+    bool new_archival = new_ntp_config.shadow_indexing_mode
+                        && model::is_archival_enabled(
+                          new_ntp_config.shadow_indexing_mode.value());
+    if (
+      old_ntp_config.is_archival_enabled() != new_archival
+      || old_ntp_config.is_read_replica_mode_enabled()
+           != new_ntp_config.read_replica) {
+        cloud_storage_changed = true;
+    }
+
+    // Pass the configuration update into the storage layer
+    co_await _raft->log().update_configuration(new_ntp_config);
+
+    // If this partition's cloud storage mode changed, rebuild the archiver.
+    // This must happen after raft update, because it reads raft's ntp_config
+    // to decide whether to construct an archiver.
+    if (cloud_storage_changed) {
+        vlog(
+          clusterlog.debug,
+          "update_configuration[{}]: updating archiver for config {}",
+          new_ntp_config,
+          _raft->ntp());
+        if (_archiver) {
+            co_await _archiver->stop();
```

Comment:
maybe it makes sense to construct or stop `remote_partition` here in case if `redpanda.remote.read` changed